### DISCUSSION TITLE
Boundary-restricted variables?

### DISCUSSION THREAD

**Discussion by stephen-dixon** [idaholab/moose#26761] [Q&A General]
Hi,

I'm interested in modelling gas diffusion in metals and wanted to implement some new surface models which would use two coupled non-linear variables - one for the concentration of gas atoms in the bulk material and one for the surface coverage fraction of a monolayer of gas atoms adsorbed to the surface of the metal. The governing equations comprise 2 pdes and describe mass balance between adsorption/desorption between an external gas volume and the surface layer, and diffusion between the surface layer and the bulk concentration, and then onward diffusion downstream in the bulk only. 

I was wondering what the best way to implement this scenario in moose would be? since I can't see anything that can act as a boundary restricted variable and, ideally, I'd prefer not to use an additional layer of cells for all external surfaces of any mesh trying to implement this surface model; spatially, the thickness of this layer would be one atom thick compared to mm or m thickness of the bulk material. Is there another way to think about this or are the extra cells unavoidable?

**Comment by GiudGiud** [comment]
Hello

There are two ways to implement boundary restricted variables. The most straightforward is to create a lowerD block on the surface using this mesh generator:
https://mooseframework.inl.gov/source/meshgenerators/LowerDBlockFromSidesetGenerator.html

then block-restrict a variable of a type that can support it (it will give you an error if not) to this block.

then you have to write kernels that make sense given the 2D support of the variable

Guillaume

**Reply by pharshaad** [reply]
Hi @GiudGiud , 

Can you please elaborate on what you mean by "kernels that make sense given the 2D support of the variable"? I'm trying to simulate an FSI problem where I solve linear elastic solid mechanics on a 3D block and also solve Reynolds thin film lubrication equations on a 2D surface. The 2D domain is created from the top surface of the solid block using the LowerDBlockFromSidesetGenerator, as you suggested. The fluid equation is the following

![equation](https://github.com/user-attachments/assets/181adb11-79e8-49e4-98fb-18a13f70d3a3)

I'm restricting the variables u_x and u_y to the solid block, u_z to the solid and fluid domains, and p (pressure) to the fluid domain. I'm using a CoupledTimeDerivative type kernel for the LHS and the MattDiffusion kernel for the RHS with h^3/12mu comming from DerivativeParsedMaterial. The variable for these kernels is p. The solid is coupled with the fluid through h and the fact that u_z exists on both domains. The fluid pressure should act as normal traction on the same surface used to define the 2D block for which I'm using a CoupledPressureBC type boundary with off diagonal jacobian pressure contributions. The setup makes sense to me, however, the results I'm getting don't seem to reflect pressure effects on the solid as if it's a one-way coupled system. I wonder if this might be due to the lower and higher dimensional coupling and if you have any insight in getting the fully coupled system working.

**Reply by GiudGiud** [reply]
Hello

I mean that not all kernels make sense when applied on a 2D domain sitting along a 3D domain. The curvature of that 2D domain likely needs to be taken into consideration in the equations for example. 
Our experiences with lower dimensional meshes are very specific. We used them for mortar contact, we also used them to describe face variables in a 3D simulation. 

I imagine if you have "thin film lubrication equations", the shape of the 2D domain, notably the curvature, matters in those equations. 

>  u_z to the solid and fluid domains,

does u_z have an equation in both domains? A variable being defined on a domain means it is being solved for DOFs on that domain. If not, if you just need a projection from high-D to low-D, then block restricting to both domains is not how to do this. 

>  I wonder if this might be due to the lower and higher dimensional coupling and if you have any insight in getting the fully coupled system working.

Did you try these instructions to make sure the problem is well-posed first?

https://mooseframework.inl.gov/moose/application_usage/failed_solves.html
